File: bound/kt/src/main/kotlin/tbdex/sdk/httpclient/Balances.kt
package tbdex.sdk.httpclient

import tbdex.sdk.TbdexException
import tbdex.sdk.resources.Balance
import tbdex.sdk.rust.getBalances as rustCoreGetBalances
import tbdex.sdk.rust.fromWeb5
import tbdex.sdk.rust.BearerDid as RustCoreBearerDid
import web5.sdk.dids.BearerDid

fun getBalances(pfiDidUri: String, bearerDid: BearerDid): List<Balance> {
    try {
        val rustCoreBalances = rustCoreGetBalances(pfiDidUri, RustCoreBearerDid.fromWeb5(bearerDid))
        return rustCoreBalances.map { Balance.fromRustCoreBalance(it) }
    } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
        throw TbdexException.fromRustCore(e)
    }
}

File: bindings/tbdex_uniffi/src/http_client/balances.rs
use crate::{errors::Result, get_rt, resources::balance::Balance};
use std::sync::{Arc, RwLock};
use web5_uniffi_wrapper::dids::bearer_did::BearerDid;

pub fn get_balances(pfi_did_uri: String, bearer_did: Arc<BearerDid>) -> Result<Vec<Arc<Balance>>> {
    let rt = get_rt()?;
    let inner_balances = rt.block_on(tbdex::http_client::balances::get_balances(
        &pfi_did_uri,
        &bearer_did.0.clone(),
    ))?;

    let balances = inner_balances
        .into_iter()
        .map(|b| Arc::new(Balance(Arc::new(RwLock::new(b)))))
        .collect();

    Ok(balances)
}
